qutip/tests/test_gates.py

```diff
@@ -31,430 +31,329 @@
+        for target in range(self.n_qubits):
+            start = qutip.tensor(random[:target] + [base] + random[target:])
+            test = gate(*angles, self.n_qubits, target) * start
+            expected = qutip.tensor(random[:target] + [applied]
+                                    + random[target:])
+            assert _infidelity(test, expected) < 1e-12
+
+    @pytest.mark.parametrize(['gate', 'n_controls'], [
+        pytest.param(gates.cnot, 1, id="cnot"),
+        pytest.param(gates.cy_gate, 1, id="cY"),
+        pytest.param(gates.cz_gate, 1, id="cZ"),
+        pytest.param(gates.cs_gate, 1, id="cS"),
+        pytest.param(gates.ct_gate, 1, id="cT"),
+        pytest.param(gates.swap, 0, id="swap"),
+        pytest.param(gates.iswap, 0, id="iswap"),
+        pytest.param(gates.sqrtswap, 0, id="sqrt(swap)"),
+        pytest.param(functools.partial(gates.molmer_sorensen, 0.5*np.pi), 0,
+                     id="Molmer-Sorensen")
+    ])
+    def test_two_qubit(self, gate, n_controls):
+        targets = [qutip.rand_ket(2) for _ in [None]*2]
+        others = [qutip.rand_ket(2) for _ in [None]*self.n_qubits]
+        reference = gate() * qutip.tensor(*targets)
+        for q1, q2 in itertools.permutations(range(self.n_qubits), 2):
+            qubits = others.copy()
+            qubits[q1], qubits[q2] = targets
+            args = [[q1, q2]] if n_controls == 0 else [q1, q2]
+            test = gate(self.n_qubits, *args) * qutip.tensor(*qubits)
+            expected = _tensor_with_entanglement(qubits, reference, [q1, q2])
+            assert _infidelity(test, expected) < 1e-12
+
+    @pytest.mark.parametrize(['gate', 'n_controls'], [
+        pytest.param(gates.fredkin, 1, id="Fredkin"),
+        pytest.param(gates.toffoli, 2, id="Toffoli"),
+        pytest.param(_make_random_three_qubit_gate(), 2, id="random"),
+    ])
+    def test_three_qubit(self, gate, n_controls):
+        targets = [qutip.rand_ket(2) for _ in [None]*3]
+        others = [qutip.rand_ket(2) for _ in [None]*self.n_qubits]
+        reference = gate() * qutip.tensor(targets)
+        for q1, q2, q3 in itertools.permutations(range(self.n_qubits), 3):
+            qubits = others.copy()
+            qubits[q1], qubits[q2], qubits[q3] = targets
+            args = [q1, [q2, q3]] if n_controls == 1 else [[q1, q2], q3]
+            test = gate(self.n_qubits, *args) * qutip.tensor(*qubits)
+            expected = _tensor_with_entanglement(qubits, reference,
+                                                 [q1, q2, q3])
+            assert _infidelity(test, expected) < 1e-12
+
+
+class Test_expand_operator:
+    # Conceptually, a lot of these tests are complete duplicates of
+    # `TestGateExpansion`, except that they explicitly target an underlying
+    # function in `qutip.qip.operations.gates` which (as of 2020-03-01) is not
+    # called in the majority of cases---it appears to be newer than the
+    # surrounding code, but the surrounding code wasn't updated.
```

Comment:
Yes, its kind of my fault. In principle, `expand_operator` can replace all the `gate_expand_DtoN` functions. We will fix it this year.

BTW, do you think `expand_operator` can somehow be included in the core qutip?  Probably during your GSoC project? It is quite general and can also be used outside `qip`.

Reply:
It'll be good to get it all updated to be uniform, though I wouldn't be super worried about it right now because everything still does work.

I won't change `expand_operator` as part of the upcoming project, because it's acting conceptually on a higher level of abstraction than what we're changing (i.e. it uses `Qobj` methods for data access, not directly mutating `fast_csr_matrix` types), but I do agree that it could be promoted up into the global namespace and out of the `qip` module.  It's effectively a helper function wrapping around `tensor`, and I think it's a really useful operation that would really ease operator construction (though the cyclic permutation option is perhaps a bit specific to be given such a prominent place).